src/main/java/org/apache/sling/feature/cpconverter/repoinit/createpath/CreatePathSegmentProcessor.java

```diff
@@ -27,52 +29,97 @@
+    public boolean processSegments() {
         for (final String part : path.getSegments()) {
-            final String platformName = PlatformNameFormat.getPlatformName(part);
-            repositoryPath = repositoryPath.concat(ConverterConstants.SLASH).concat(platformName);
+            repositoryPath = processSegment(part);
+        }
+        return foundType;
+    }
 
-            boolean segmentAdded = false;
-            //loop all package assemblers and check if .content.xml is defined
-            for (VaultPackageAssembler packageAssembler : packageAssemblers) {
-                File currentContent = packageAssembler.getFileEntry(repositoryPath.concat(ConverterConstants.SLASH).concat(DOT_CONTENT_XML));
-                if (currentContent.exists() && currentContent.isFile()) {
-                    //add segment if jcr:primaryType is defined.
-                    segmentAdded = addSegment(cp, part, currentContent);
-                    if (segmentAdded) {
-                        foundType = true;
-                        break;
+    @NotNull
+    private String processSegment(String part) {
+        final String platformName = PlatformNameFormat.getPlatformName(part);
+        repositoryPath = repositoryPath.concat(ConverterConstants.SLASH).concat(platformName);
+        
+        //loop all package assemblers and check if .content.xml is defined
+        collectTypeDataForSegment();
+        addSegment(part);
+        
+        return repositoryPath;
+    }
+    
+    private void addSegment(String part){
+        //add segment if jcr:primaryType is defined.
+        if(primaryTypeMap.containsKey(repositoryPath)){
+            cp.addSegment(
+                    part, 
+                    primaryTypeMap.get(repositoryPath), 
+                    mixinTypeMap.get(repositoryPath)
+            );
+        }else{
+            cp.addSegment(part, null);
+        }
+    }
+
+    private void collectTypeDataForSegment() {
+        for (VaultPackageAssembler packageAssembler : packageAssemblers) {
+            
+            if(primaryTypeMap.containsKey(repositoryPath)){
+                boolean merge = true;
+                for(PathFilterSet set: packageAssembler.getFilter().getFilterSets()){
+                    if(set.covers(repositoryPath) && (set.getImportMode() != ImportMode.MERGE && set.getImportMode() != ImportMode.MERGE_PROPERTIES)){
+                        //found a path with a mode other than merge, proceed to replace the type definitions
+                        merge = false;
+                    }else if(set.covers(repositoryPath)){
+                        // merge is flipped to true again by another filter defined after the previous ones
+                        merge = true;
                     }
                 }
+                if(merge){
+                    // we already got the path defined by an earlier package. only proceed to replace if the filter is not merge.
+                    continue;
```

Comment:
wait-what.... ? ah.... that's the outer for-loop..... 
that code looks quite brittle IMHO. plese refactor the code such that you don't need the 'continue' and make the code easy to understand/read/maintain.